### PR TITLE
librbd: initial set of changes to migration API for instant-restore

### PR DESCRIPTION
**Pull request by dillaman**
Start the process of allowing a JSON-encoded migration source-spec string to be provided. This implies that migration should always attempt to first open the destination image for execute/commit/abort actions. This also includes an initial implementation of a raw, file-based image source IO path handler.

### CODE REVIEW

File: src/librbd/migration/ImageDispatch.cc [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+  *dispatch_result = io::DISPATCH_RESULT_COMPLETE;
+
+  uint64_t clip_length = 0;
+  for (auto [_, image_length] : image_extents) {
+    clip_length += image_length;
+  }
+
+  aio_comp->read_result = std::move(read_result);
+  aio_comp->read_result.set_clip_length(clip_length);
```
Comment (trociny) [inline src/librbd/migration/ImageDispatch.cc:54]: @dillaman needs update after #37727 is merged?

Comment (dillaman) [reply src/librbd/migration/ImageDispatch.cc]: Done

File: src/cls/rbd/cls_rbd_types.h [hunk 2]
```diff
@@ -983,10 +985,11 @@ struct MigrationSpec {
   inline bool operator==(const MigrationSpec& ms) const {
     return header_type == ms.header_type && pool_id == ms.pool_id &&
       pool_namespace == ms.pool_namespace && image_name == ms.image_name &&
-      image_id == ms.image_id && snap_seqs == ms.snap_seqs &&
-      overlap == ms.overlap && flatten == ms.flatten &&
-      mirroring == ms.mirroring && mirror_image_mode == ms.mirror_image_mode &&
-      state == ms.state && state_description == ms.state_description;
+      image_id == ms.image_id && source_spec == ms.source_spec &&
```
Comment (trociny) [inline src/cls/rbd/cls_rbd_types.h:988]: nit: just wondering, wouldn't it be more correct to compare the source spec just after comparing the header type?

It wouldn't make any difference in real life though I suppose...

File: src/librbd/migration/NativeFormat.cc [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+template <typename I>
+void NativeFormat<I>::read(uint64_t snap_id, io::Extents&& image_extents,
+                           int op_flags, int read_flags, bufferlist* data,
+                           const ZTracer::Trace &parent_trace,
+                           Context* on_finish) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << "snap_id=" << snap_id << ", "
+                 << "image_extents=" << image_extents << dendl;
+  io::ReadResult read_result{data};
```
Comment (trociny) [inline src/librbd/migration/NativeFormat.cc:126]: @dillaman Do I understand right that we don't support sparse read here?
And the same for `RawFormat`?

Comment (trociny) [reply src/librbd/migration/NativeFormat.cc:126]: But probably not a big deal, if it is not going to be used by background migration (deep copy).

**Comment by trociny** [comment]
And just wondering, what was the reason to use json for the migration source spec? It seems that "ceph native encoding" could work here too, similarly how it is used e.g. for encoding different journal event types.

**Comment by dillaman** [comment]
> And just wondering, what was the reason to use json for the migration source spec? It seems that "ceph native encoding" could work here too, similarly how it is used e.g. for encoding different journal event types.

We want someone to be able to define it on the CLI without using a hexeditor, so once we have a JSON/XML format, we might was well just keep it in the single format instead of creating yet another format (that would also need to be extensible to permit new formats, streams, etc).

File: src/librbd/migration/NativeFormat.cc (hunk 3 above)
Comment (dillaman) [reply src/librbd/migration/NativeFormat.cc:126]: Good point, I'll switch the API to just use a `ReadResult` to hide the underlying type -- and to allow `NativeFormat` to directly pass it and continue to utilize sparse bufferlists.

**Comment by trociny** [comment]
> We want someone to be able to define it on the CLI without using a hexeditor, so once we have a JSON/XML format, we might was well just keep it in the single format instead of creating yet another format (that would also need to be extensible to permit new formats, streams, etc).

We could make the CLI/API convert the user format to a native ceph encoding (Anyway, I suppose we might want to validate somehow the user input before storing it in the image migration header?) And users can't extent a format until it is supported by librbd, so this extension could be added to the native encoding.

But I don't have a strong opinion here. Just ignore if you like your way.

**Comment by dillaman** [comment]
If we had an intermediate binary format, the OSDs (cls_rbd methods) would still need to treat the `source_spec` as binary blob and not attempt to decode it or else adding new client-side only features would require OSD upgrades. On the client-side, the next PR adds basic validation (valid JSON, can properly open the source before creating the destination image w/ the embedded JSON). This basic validation could probably be extended to support fail or warn on unknown attributes (i.e. don't let an image be created with attributes that are not used / understood) or warn on unused/unknown arguments on open.  

The journal was all binary encoded for performance/space efficiency reasons which I don't think applies here. I think at the end of it all, JSON can provide the exact same set of forward and backward protection as native encoding w/o all the extra boilerplate code. I would think at best you just a few bytes of space savings here and there since you wouldn't have to encode key names and non-text values can be binary encoded. Just my opinion.

**Review by trociny — APPROVED** [review]
LGTM

**Comment by trociny** [comment]
@dillaman  There are test failures [1], cli and python, and valgrind reported a memory leak in deep copy [2].

[1] https://pulpito.ceph.com/trociny-2020-10-23_10:42:30-rbd-wip-mgolub-testing-distro-basic-smithi/
[2] http://qa-proxy.ceph.com/teuthology/trociny-2020-10-23_10:42:30-rbd-wip-mgolub-testing-distro-basic-smithi/5551482/teuthology.log

**Comment by dillaman** [comment]
@trociny Thanks. The first two (cli and python) were due to typos during rebases and have been corrected. I'll fix the memory leak tomorrow as a new appended commit.

**Comment by dillaman** [comment]
https://pulpito.ceph.com/jdillaman-2020-10-26_10:08:16-rbd-wip-jd-testing-distro-basic-smithi/

File: src/librbd/migration/ImageDispatch.cc [hunk 4]
```diff
@@ -44,9 +44,9 @@ bool ImageDispatch<I>::read(
   ldout(cct, 20) << dendl;
 
   *dispatch_result = io::DISPATCH_RESULT_COMPLETE;
-  m_format->read(aio_comp, io_context->read_snap().value_or(CEPH_NOSNAP),
-                 std::move(image_extents), std::move(read_result), op_flags,
-                 read_flags, parent_trace);
+  return m_format->read(aio_comp, io_context->read_snap().value_or(CEPH_NOSNAP),
+                        std::move(image_extents), std::move(read_result),
+                        op_flags, read_flags, parent_trace);
   return true;
```
Comment (trociny) [inline src/librbd/migration/ImageDispatch.cc:50]: nit: forgot to remove this line?

**Comment by trociny** [comment]
@dillaman Jenkins failed on assertion inTestLibRBD.FlushEmptyOpsOnExternalSnapshot, but I suspect it is related to my recent changes to the watcher. I am going to investigate this tomorrow (if the cause is not obvious for you).

**Comment by dillaman** [comment]
> @dillaman Jenkins failed on assertion inTestLibRBD.FlushEmptyOpsOnExternalSnapshot, but I suspect it is related to my recent changes to the watcher. I am going to investigate this tomorrow (if the cause is not obvious for you).

Yeah, it's an issue related to the watcher changes. Looks like the `SafeTimer` isn't actually configured to be safe (`m_safe_timer = new SafeTimer(cct, m_lock, false)`) so there is a possibility for the timer to fire but not get invoked w/ `m_lock` locked and therefore race w/ any of the `cancel_event` calls.

**Comment by dillaman** [comment]
jenkins test make check

**Comment by trociny** [comment]
> Yeah, it's an issue related to the watcher changes. Looks like the SafeTimer isn't actually configured to be safe (m_safe_timer = new SafeTimer(cct, m_lock, false)) so there is a possibility for the timer to fire but not get invoked w/ m_lock locked and therefore race w/ any of the cancel_event calls.

pushed #37880 for this.

**Comment by trociny** [comment]
> https://pulpito.ceph.com/jdillaman-2020-10-26_10:08:16-rbd-wip-jd-testing-distro-basic-smithi/
> https://pulpito.ceph.com/jdillaman-2020-10-26_10:23:51-rbd-wip-jd-testing-distro-basic-smithi/

@dillaman I missed this before merging (the job was still running when I was looking and later I forgot to check its status before merging), but there was one failure [1],  though I have no idea if it is related. If I interpret the logs correctly It seems like the qemu got stuck?

Just wanted to let you know in case you had not seen this and would be interested.

[1] http://qa-proxy.ceph.com/teuthology/jdillaman-2020-10-26_10:23:51-rbd-wip-jd-testing-distro-basic-smithi/5560517/teuthology.log

**Comment by dillaman** [comment]
The `dynamic_features_no_cache` has been getting randomly stuck for a while. I tried to look into it a couple weeks ago by increasing the log levels but then it failed to repeat. I'll open a ticket against it so that I don't forget about it.

**Pull request merged**